### PR TITLE
New backend API

### LINKED ISSUE #6: Implement new spec for frontend <~> backend API
Rationale: currently, the API only returns requirement ID and title in nodes/edges format, but (1) more metadata is needed (such as qualifer) and (2) nodes/edges format is not very visualization library agnostic.

Therefore, the new spec should be as follows, with the data represented as an Object with Notion IDs as fields and metadata as contents:

```json
{
  "ff5892a7-b5c9-4a25-97bd-663b2cff5bd7": {
    "created-by": "khang nguyen",
    "parent-id": "b7985ca8-94f4-4a30-ae61-8af2126019bd",
    "last-edited": "2023-02-20T17:51:00.000Z",
    "qualifier": "SHALL",
    "collection": "Tree visualizer",
    "type": "Functional",
    "verification-plan": "", // unsure how to represent this
    "verification-method": "", // unsure how to represent this
    "system": "Tree Visualizer",
    "rationale": "To allow UTAT Space Systems members to graphically view the relationships between entries",
    "trades": "", // unsure how to represent this
    "last-edited-by": "David",
    "stakeholder": "Systems",
    "mission": "FINCH",
    "description": "ADCS shall consistently know FINCH’s location relative to the SUN and Earth Position",
    "title": "FINCH-ADCS-AttitudeDetermination",
    "url": "https://www.notion.so/FINCH-Power-Batteries-74331bc1129c417c850a0a5ca0deb759"
  },

[...]

}
```

### PR DESCRIPTION
**Pull request by jdtech3**
Resolves #6 with new backend API of form:

```json
 {
  "ff5892a7-b5c9-4a25-97bd-663b2cff5bd7": {
    "created-by": "khang nguyen",
    "parent-id": "b7985ca8-94f4-4a30-ae61-8af2126019bd",
    "last-edited": "2023-02-20T17:51:00.000Z",
    "qualifier": "SHALL",
    "collection": [
      "Volume Allocation & Placement"
    ],
    "verification-plan": [
      "TVAC Testing Plan"
    ],
    "system": "Tree Visualizer",
    "rationale": "To allow UTAT Space Systems members to graphically view the relationships between entries",
    "trades": [
      "Spacecraft Bus Stack Order Selection"
    ],
    "last-edited-by": "David",
    "stakeholder": [
      "Mechanical",
      "Power",
      "Thermal"
    ],
    "mission": "FINCH",
    "description": "ADCS shall consistently know FINCH’s location relative to the SUN and Earth Position",
    "title": "FINCH-ADCS-AttitudeDetermination",
    "url": "https://www.notion.so/FINCH-Power-Batteries-74331bc1129c417c850a0a5ca0deb759"
  },

[...]

}
```

_Note: this is slightly different from description in #6 due to Requirements DB changes._

### CODE REVIEW

File: server/test/mock-notion-api.js [hunk 1]
```diff
@@ -0,0 +1,719 @@
...
+          "🍄 Type": {
+            id: "Z~z%5D",
+            type: "relation",
+            relation: [
+              {
+                id: "d484af0c-f3f9-41c9-8afc-543693ec608d",
+              },
+            ],
+            has_more: false,
+          },
```
Comment (DM1122) [inline server/test/mock-notion-api.js:100]: `Type` is no longer a property of the database, Should this be removed?

**Review by DM1122 — CHANGES_REQUESTED** [review]
This is looking solid, great work @jdtech3. I left a few comments about how we can align with the [dependecy injection](https://en.wikipedia.org/wiki/Dependency_injection#:~:text=In%20software%20engineering%2C%20dependency%20injection,leading%20to%20loosely%20coupled%20programs.) pattern more closely. The advantage of doing so being that it ensures your objects or functions which want to use a given service do not have to know how to construct those services.

File: server/src/routes/index.js [hunk 2]
```diff
@@ -1,20 +1,19 @@
...
   let edges_json = JSON.stringify(data.edges);
   res.render('index', { title: 'TreeVisualizer' , nodes: nodes_json, edges: edges_json});
 });
 
 // GET : JSON API
 router.get('/api/requirements', async function(req, res, next) {
-  let data = await notionService.getRequirementsData();
+  let data = await notionService.getRequirements('api');
   res.json(data);
 });
```
[the remark this answers is not in the document]
Comment (jdtech3) [reply server/src/routes/index.js:18]: We do not, no. Honestly it's probably not great practice, but the legacy demo functionality is serving as an easy sanity test to see whether data is being fetched from the Notion API correctly—JSON is hard to read, but if vis graph generates, data is probably fine.

Having said that, no real need, should I remove?

File: server/src/services/notion.service.js [hunk 3]
```diff
@@ -1,39 +1,119 @@
...
+  return await _getDBTitles(getEnv('NOTION_MISSIONS_DB_ID'));
+}
+
+const _parseRequirementsForAPI = async function(raw_data) {
+  let qualifiers = await getRequirementQualifiers();
+  let test_plans = await getTestPlans();
+  let systems = await getSystems();
+  let trades = await getTrades();
+  let teams = await getTeams();
+  let missions = await getMissions();
```
[the remark this answers is not in the document]
Comment (jdtech3) [reply server/src/services/notion.service.js]: Good idea, resolved.

File: server/src/services/notion.service.js [hunk 4]
```diff
@@ -1,39 +1,119 @@
...
 
-const getRequirementsData = async function() {
-  let all = []
+  return new notion.Client({
+    auth: auth_token,
+  });
+}
+
+const _getNotionDB = async function(database_id, notion_client = undefined) {
+  if (notion_client === undefined) notion_client = newNotionClient();
```
[the remark this answers is not in the document]
Comment (jdtech3) [reply server/src/services/notion.service.js:13]: Definitely non-ideal but to follow the pattern would require all of the `_get<DB>` commands to either take in a `notion_client` or a `auth_token`, which I decided against as it'd result in a lot more repeated code.

File: server/test/mock-notion-api.js (hunk 1 above)
Comment (jdtech3) [reply server/test/mock-notion-api.js]: Oops, updated!

File: server/src/routes/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (DM1122) [reply server/src/routes/index.js:18]: Gotcha, your call to make, if its a quality of life feature we can keep both

File: server/src/services/notion.service.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (DM1122) [reply server/src/services/notion.service.js:13]: I see, we can revisit this at a later point to find a way to structure this code so it complies with DI without needing to duplicate code. Closing these as resolved

**Review by DM1122 — APPROVED** [review]

**Pull request merged**